### ISSUE TITLE
Typescript errors for context object in Apollo gateway

### ISSUE THREAD

**Issue by toinevk** [apollographql/apollo-server#3307]
Package Versions: 
`@apollo-gateway: 0.10.4`

Using the code below (from: [https://www.apollographql.com/docs/apollo-server/federation/implementing/]) I wouldn't expect to see any errors when using typescript. 

```
const gateway = new ApolloGateway({
  serviceList: [
    { name: 'products', url: 'http://localhost:4001' },
    // other services
  ],
  buildService({ name, url }) {
    return new RemoteGraphQLDataSource({
      url,
      willSendRequest({ request, context }) {
        // pass the user's id from the context to underlying services
        // as a header called `user-id`
        request.http.headers.set('user-id', context.userId);
      },
    });
  },
});
```

However, typescript warns me that "Property 'userId' does not exist on type 'TContext'."

Looking at the underlying code it appears that willSendRequest should accept any key value pair with a string but that doesn't appear to be the case. Although Im not sure where the error could be.

Sorry if I have missed anything, Im still learning some of the advanced methods in typescript.

**Comment by hnrklssn** [comment]
I have the same issue when using the example with `willSendResponse` listed in https://www.apollographql.com/docs/apollo-server/federation/implementing/#customizing-outgoing-responses

**Comment by migueloller** [comment]
Adding to this issue, if a type of `{ request: GraphQLRequest, context: Context }` is added, TypeScript will complain that the provided `Context` isn't compatible with `TContext`.

**Comment by rylanc** [comment]
Any update on this? I'm also having trouble adding types to `willSendRequest`.

**Comment by LucienLee** [comment]
It looks the author misunderstands the generic type in the class. It should pass generic type from class to it's method, but in [code](https://github.com/apollographql/apollo-server/blob/master/packages/apollo-gateway/src/datasources/RemoteGraphQLDataSource.ts) it only set generic type on the method.

The correct one would be:

```ts
export class RemoteGraphQLDataSource<TContext> implements GraphQLDataSource { // <- pass generic from class
  public willSendRequest?<TContext>( // <- so we could pass our own context type here
    requestContext: Pick<
      GraphQLRequestContext<TContext>,
      'request' | 'context'
    >,
  ): ValueOrPromise<void>;
}
```

**Comment by abernix** [comment]
Should be fixed by #3865.  Please report back if this isn't the case!

**Comment by jaymarkjairi** [comment]
> I ran into this issue today trying to clean up some of my gateway code. I've fixed the Typescript generics typing in PR #3865. In the meantime, there is a workaround I've been using in my gateway currently:
> 
> ```ts
> import { ApolloGateway, RemoteGraphQLDataSource } from "@apollo/gateway";
> import { GraphQLRequest } from "apollo-server-core";
> 
> interface IContext {
>   user?: {
>     id: string;
>   };
> }
> 
> const gateway = new ApolloGateway({
>   buildService({ url }) {
>     return new RemoveGraphQLDataSource<IContext>({
>       url,
>       willSendRequest({ request, context }: { request: GraphQLRequest, context: IContext }) {
>         // "context" will now be typed to IContext
>       }
>     });
>   }
> });
> ```

this is my error in applying this

```
Type '({ request, context }: { request: GraphQLRequest; context: ICustomContext; }) => void' is not assignable to type '(options: GraphQLDataSourceProcessOptions<ICustomContext>) => ValueOrPromise<void>'.
  Types of parameters '__0' and 'options' are incompatible.
    Type 'GraphQLDataSourceProcessOptions<ICustomContext>' is not assignable to type '{ request: GraphQLRequest; context: ICustomContext; }'.
      Type '{ request: GraphQLRequest; } & { kind: GraphQLDataSourceRequestKind.HEALTH_CHECK | GraphQLDataSourceRequestKind.LOADING_SCHEMA; context: {}; }' is not assignable to type '{ request: GraphQLRequest; context: ICustomContext; }'.
        Types of property 'context' are incompatible.
          Type '{}' is missing the following properties from type 'ICustomContext': req, rests(2322)
```

**Comment by glasser** [comment]
@jaymarkjairi This would be better filed as a new issue on https://github.com/apollographql/federation .

That said, the typing for the `process` method on `GraphQLDataSource changed in `@apollo/gateway@0.36.0`. See https://github.com/apollographql/federation/pull/870/files#diff-89ab698c91f23713754f05923839b88aebd5178365ea80590ac7d964a46b90a7 for details.
